yarn-project/types/src/l1_to_l2_message.ts

```diff
@@ -17,6 +17,79 @@ export interface L1ToL2MessageSource {
   getPendingL1ToL2Messages(take: number): Promise<L1ToL2Message[]>;
 }
 
+/**
+ * A simple in-memory implementation of an L1 to L2 message store
+ * that handles message duplication.
+ */
+export class L1ToL2MessageStore {
+  /**
+   * A map containing the message key to the corresponding L1 to L2
+   * messages (and the number of times the message has been seen).
+   */
+  private store: Map<Fr, L1ToL2MessageAndCount> = new Map();
+
+  constructor() {}
+
+  addMessage(entryKey: Fr, msg: L1ToL2Message) {
+    if (this.store.has(entryKey)) {
+      this.store.get(entryKey)!.count++;
+    } else {
+      this.store.set(entryKey, { message: msg, count: 1 });
+    }
+  }
+
+  combineWith(other: L1ToL2MessageStore) {
+    other.store.forEach((value, key) => {
+      if (this.store.has(key)) {
+        this.store.get(key)!.count += value.count;
+      } else {
+        this.store.set(key, value);
+      }
+    });
+  }
+
+  removeMessage(entryKey: Fr) {
+    if (!this.store.has(entryKey)) {
```

Comment:
Feels like this should be an error? If it is consumed and you never had it pending, something is not good.

I'm thinking that you did this to satisfy tests that run into issues because of "0-messages"?